platform/security/platform-security-core-api/src/main/java/ddf/security/samlp/SimpleSign.java

```diff
@@ -188,17 +196,35 @@ public void signUriString(String queryParams, UriBuilder uriBuilder) throws Sign
     return signature;
   }
 
-  private String getSignatureAlgorithm(X509Certificate certificate) {
-    String sigAlgo = SSOConstants.RSA_SHA1;
-    String pubKeyAlgo = certificate.getPublicKey().getAlgorithm();
+  private java.security.Signature getSignature(X509Certificate certificate)
+      throws SignatureException {
+    String jceSigAlgo = certificate.getSigAlgName();
 
-    if (pubKeyAlgo.equalsIgnoreCase("DSA")) {
-      sigAlgo = SSOConstants.DSA_SHA1;
+    java.security.Signature signature;
+    try {
+      if ("DSA".equalsIgnoreCase(certificate.getPublicKey().getAlgorithm())) {
+        signature = java.security.Signature.getInstance(jceSigAlgo, "BC");
+      } else {
+        signature = java.security.Signature.getInstance(jceSigAlgo);
+      }
+    } catch (NoSuchAlgorithmException | NoSuchProviderException e) {
+      throw new SignatureException(e);
     }
+    return signature;
+  }
 
-    LOGGER.debug("Using Signature algorithm {}", sigAlgo);
+  private String getSignatureAlgorithmURI(X509Certificate certificate) {
+    String sigAlgoUri = RFC5698_DSSC_OID_TO_URI_MAP.get(certificate.getSigAlgOID());
 
-    return sigAlgo;
+    if (sigAlgoUri == null) {
+      sigAlgoUri =
+          ("DSA".equalsIgnoreCase(certificate.getPublicKey().getAlgorithm()))
+              ? WSS4JConstants.SHA384
```

Comment:
❓ Where did `SHA384` come from?

Reply:
@stustison and I discussed and decided to use the more secure SHA384 as our fallback hashing algorithm for DSA.